File: 1510. Stone Game IV.js
// https://leetcode.com/problems/stone-game-iv/

/**
 * @param {number} n
 * @return {boolean}
 */
var winnerSquareGame = function(n) {

    const memo = new Array(n+1).fill(0).map(() => []);

    const aliceAlwaysWins = (n, aliceTurn) => {
        const root = Math.sqrt(n); // if the root is zero
        if(Math.floor(root) === root) return aliceTurn; // the current player wins

        if(memo[n][aliceTurn] !== undefined) return memo[n][aliceTurn]; //

        if(aliceTurn){ // if it's alice's turn, alice always wins if
            for(let i=1; i<=root; i++){ // there's a move she can make that causes her to win
                if(aliceAlwaysWins(n - i*i, false)) return memo[n][aliceTurn] = true; // with n-i*i stones
            }
            return memo[n][aliceTurn] = false;
        }

        // And if its bob's turn
        for(let i=1; i<=root; i++){ // if there's a single move bob can make that makes alice not win
            if(!aliceAlwaysWins(n - i*i, true)) return memo[n][aliceTurn] = false; // then alice doesn't always win
        }
        return memo[n][aliceTurn] = true // but if there's no such move, bob wins
    };

    return aliceAlwaysWins(n, true);
};

File: 1510. Stone Game IV.go
// https://leetcode.com/problems/stone-game-iv/

package leetcode

import "math"

func winnerSquareGame(n int) bool {
	has, cached := newBitset(n+1), newBitset(n+1)

	var currentPlayerWins func(int) bool
	currentPlayerWins = func(n int) bool {
		switch {
		case n == 0:
			return false
		case has.Get(n):
			return cached.Get(n)
		default:
			has.Set(n)
			root := int(math.Sqrt(float64(n)))
			for i := 1; i <= root; i++ {
				if !currentPlayerWins(n - i*i) {
					cached.Set(n)
					return true
				}
			}
			return false
		}
	}

	return currentPlayerWins(n)
}

// use a nifty bitset for speed
const wordSize = 64

type bitset struct {
	data []uint64
}

func newBitset(size int) *bitset {
	return &bitset{data: make([]uint64, (size/wordSize)+1)}
}

func (b *bitset) Get(x int) bool {
	return b.data[x/wordSize]&(1<<(x%wordSize)) != 0
}

func (b *bitset) Set(x int) {
	b.data[x/wordSize] |= (1 << (x % wordSize))
}
